implementing padding

owendevita opened the pull request:
This patch implements the random padding surrounding the BSON packet. The one included unit test ensures a proper wrapping and unwrapping of our packet. Future needs may warrant further unit testing for invalid data.

Signed-off-by: Owen De Vita <owen.n.devita@gmail.com>

owendevita committed:
changes

owendevita committed:
Update padding.rs

amyipdev commented on src/padding.rs:
```diff
@@ -1,12 +1,130 @@
+    }
+
+    // the revealed packet is just the bytes between the opening bracket and closing bracket
+    let revealed_pkt = pkt_vec[bson_front_index as usize..bson_back_index as usize].to_vec();
```
Why cast usize into i32 and then back into usize?

amyipdev commented on src/padding.rs:
```diff
@@ -1,12 +1,130 @@
+    let mut bson_back_index = -1;
+
+    // loop through the packet and find the index of the beginning of the BSON data.
+    for (index, &element) in pkt_vec.iter().take(33).enumerate() {
```
This can panic if the input is small enough.

amyipdev commented on src/padding.rs:
```diff
@@ -1,12 +1,130 @@
+    }
+
+    // loop through the packet backwards to find the end of the BSON data
+    for (rev_index, &element) in pkt_vec.iter().rev().take(33).enumerate() {
```
This can also panic.

amyipdev commented on src/padding.rs:
```diff
@@ -1,12 +1,130 @@
+/// if there is an error during the extraction of BSON data.
 fn reveal(pkt: &[u8]) -> Result<Vec<u8>, std::io::Error> {
-    unimplemented!()
+    let pkt_vec = pkt.to_vec();
```
Is this necessary? Could you act just on the slice instead of going into a vec again?

amyipdev commented on src/padding.rs:
```diff
@@ -1,12 +1,130 @@
+    back_pad.insert(0, '}' as u8);
+
+    // define a new vector to add the front pad, the packet, and finally the back pad
+    let mut encapsulated_pkt = Vec::new();
```
This is something that should probably be done on aes too - but if we are going to use with_capacity somewhere, why not use it everywhere?

amyipdev commented on src/padding.rs:
```diff
@@ -1,12 +1,130 @@
+/// * `exclusion` - The character to exclude from the generated bytes.
+/// # Returns
+/// A vector of random bytes with the defined length, excluding the specified character.
+fn generate_random_bytes<R: Rng>(rng: &mut R, count: u32, exclusion: char) -> Vec<u8> {
```
This can probably be inlined, it can be done fairly simply. Generics here really aren't necessary and this solution is overkill, especially since we are drawing 256 even if we only need 2 (which on low entropy systems is a huge issue perf wise)

amyipdev reviewed (changes requested):
Much better, still some work to do. Feel free to just push commits onto your existing branch

owendevita committed:
Changes Resolved + Readability

Applied changes including:
Inlining the generation of random bytes
Avoiding panic on .take()
Working directly on the slice
Using with_capacity

Improved readability with constants for future updates if needed.

amyipdev commented on src/padding.rs:
```diff
@@ -1,12 +1,115 @@
+    for (index, &element) in pkt.iter().enumerate().take(33) {
+        if element == b'{' {
+            // if the brace is not within 2-32 (inclusive) bytes, the padding is not valid
+            if index < 2 || index > 32 {
```
minor one, but shouldnt this use `MIN_BYTES_PADDING`/`MAX_BYTES_PADDING` now that those are defined?

amyipdev commented on src/padding.rs:
```diff
@@ -1,12 +1,115 @@
+    // loop through the packet backwards to find the end of the BSON data
+    for (rev_index, &element) in pkt.iter().rev().enumerate().take(33) {
+        if element == b'}' {
+            if rev_index < 2 || rev_index > 32 {
```
See previous comment

amyipdev reviewed (changes requested):
Looks almost good, just two minor things (really one in two places) and then we're ready to merge!

owendevita committed:
Implement Constant Variables

Improve the readability of the code by implementing the constant variables more consistently.

amyipdev reviewed (approved):
LGTM!

Reviewed-by: Amy Parker <amy@amyip.net>
